### PR TITLE
URL Resolver

### PR DESCRIPTION
**Pull request by tjwiebell**

### CODE REVIEW

File: src/Resolver/Template.php [hunk 1]
```diff
@@ -53,11 +53,11 @@ public function resolve($definition)
         $renderData     = [];
 
         if ($definition->has('provide')) {
-            foreach ($definition->get('provide') as $index => $definition) {
```
Comment (zetlen) [inline src/Resolver/Template.php]: Is this part of the PR functionality or just a little naming cleanup you did along the way?

File: src/Resolver/Url.php [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+
+        if ($definition->has('hash')) {
+            $uri->setFragment(str_replace('#', '', $this->getIterator()->get('hash', $definition)));
+        }
+
+        $returnUrl = $uri->toString();
+
+        return $uri->getHost() === self::FAKE_BASE_HOST
+            ? str_replace(self::FAKE_BASE_URL, '', $returnUrl)
+            : $returnUrl;
```
Comment (zetlen) [inline src/Resolver/Url.php]: In the spec, these values are supposed to be usable either as strings via coercion, or as objects with URL part properties: https://github.com/magento-research/pwa-studio/tree/develop/packages/upward-spec#urlresolver-example (last line of the section).

Looks like this returns a string. I don't see a test for the property lookup behavior. If it's hard to implement, we could revise the spec to lose that requirement (though I'd prefer it stay in). What do you think?

File: src/Resolver/Template.php (hunk 1 above)
Comment (tjwiebell) [reply src/Resolver/Template.php]: Yeah, two birds one PR to close #62 as well.

File: src/Resolver/Url.php (hunk 2 above)
Comment (tjwiebell) [reply src/Resolver/Url.php]: Not intentional, just an oversight. That seems simple enough, though the return type in `upward-js` appears to be a node URL class, or an object with only a subset of those properties. Should I follow that pattern in this implementation?

**Comment by tjwiebell** [comment]
Refresh cloud instance with this change to UPWARD and latest copy of Venia. Will run regression tests, and will likely just merge on pass with @dpatil-magento approval. We'll be running a full in depth QA once the resolver is implemented into Venia in scope of #1166.

https://master-7rqtwti-xjmpcdjpe6kzm.us-4.magentosite.cloud/

**Pull request merged**